File: nosql-vector-search-typescript/src/vector-search.ts
 import type { Container, SqlQuerySpec } from '@azure/cosmos';
 import path from 'path';
 import { readFileReturnJson, getClientsPasswordless, validateFieldName, insertData, printSearchResults, getQueryActivityId } from './utils.js';

// ESM specific features - create __dirname equivalent
import { fileURLToPath } from "node:url";
import { dirname } from "node:path";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

type VectorAlgorithm = 'diskann' | 'quantizedflat';

interface AlgorithmConfig {
    containerName: string;
    algorithmName: string;
}

const algorithmConfigs: Record<VectorAlgorithm, AlgorithmConfig> = {
    diskann: {
        containerName: 'hotels_diskann',
        algorithmName: 'DiskANN'
    },
    quantizedflat: {
        containerName: 'hotels_quantizedflat',
        algorithmName: 'QuantizedFlat'
    }
};

const ALGORITHM_ORDER: VectorAlgorithm[] = ['diskann', 'quantizedflat'];
const VALID_DISTANCE_FUNCTIONS = ['cosine', 'euclidean', 'dotproduct'] as const;
const METRIC_LABELS: Record<(typeof VALID_DISTANCE_FUNCTIONS)[number], string> = {
    cosine: 'COS',
    euclidean: 'L2',
    dotproduct: 'IP'
};

interface ComparisonRow {
    algorithmName: string;
    metric: (typeof VALID_DISTANCE_FUNCTIONS)[number];
    results: any[];
}

const config = {
    query: "quintessential lodging near running trails, eateries, retail",
    dbName: "Hotels",
    algorithm: (process.env.VECTOR_ALGORITHM || 'diskann').trim().toLowerCase() as VectorAlgorithm,
    dataFile: process.env.DATA_FILE_WITH_VECTORS!,
    embeddedField: process.env.EMBEDDED_FIELD!,
    embeddingDimensions: parseInt(process.env.EMBEDDING_DIMENSIONS! || process.env.VECTOR_EMBEDDING_DIMENSIONS || '1536', 10),
    deployment: process.env.AZURE_OPENAI_EMBEDDING_MODEL!,
    distanceFunction: (process.env.VECTOR_DISTANCE_FUNCTION || 'cosine').toLowerCase(),
    compareMetrics: (process.env.COMPARE_DISTANCE_METRICS || 'false').toLowerCase() === 'true',
};

function buildVectorQuerySpec(
    embedding: number[],
    safeEmbeddedField: string,
    distanceFunction: string,
    topK: number,
    scoreAlias: string
): SqlQuerySpec {
    return {
        query: `SELECT TOP ${topK} c.HotelName, c.Description, c.Rating, VectorDistance(c.${safeEmbeddedField}, @embedding, false, {"distanceFunction": "${distanceFunction}"}) AS ${scoreAlias} FROM c ORDER BY VectorDistance(c.${safeEmbeddedField}, @embedding, false, {"distanceFunction": "${distanceFunction}"})`,
        parameters: [
            { name: '@embedding', value: embedding }
        ]
    };
}

function truncateHotelName(name: string): string {
    return name.length > 20 ? `${name.slice(0, 20)}..` : name;
}

function printComparisonTable(rows: ComparisonRow[]) {
    console.log('\n| Algorithm     | Metric | Top 1 Result            | Score  | Top 2 Result            | Score  |');
    console.log('|---------------|--------|-------------------------|--------|-------------------------|--------|');

    for (const row of rows) {
        const top1 = row.results[0];
        const top2 = row.results[1];
        const top1Name = top1?.HotelName ? truncateHotelName(top1.HotelName) : 'N/A';
        const top2Name = top2?.HotelName ? truncateHotelName(top2.HotelName) : 'N/A';
        const top1Score = typeof top1?.SimilarityScore === 'number' ? top1.SimilarityScore.toFixed(4) : 'N/A';
        const top2Score = typeof top2?.SimilarityScore === 'number' ? top2.SimilarityScore.toFixed(4) : 'N/A';

        console.log(
            `| ${row.algorithmName.padEnd(13)} | ${METRIC_LABELS[row.metric].padEnd(6)} | ${top1Name.padEnd(24)} | ${top1Score.padStart(6)} | ${top2Name.padEnd(24)} | ${top2Score.padStart(6)} |`
        );
    }

    console.log('\n====================================================================================================');
    console.log('Summary: Compared 2 algorithms x 3 metrics = 6 combinations');
    console.log('====================================================================================================');
}

async function runSingleMetricQuery(container: Container, embedding: number[], safeEmbeddedField: string, distanceFunction: string) {
    const querySpec = buildVectorQuerySpec(embedding, safeEmbeddedField, distanceFunction, 5, 'SimilarityScore');

    console.log('\n--- Executing Vector Search Query ---');
    console.log('Query:', querySpec.query);
    console.log('Parameters: @embedding (vector with', embedding.length, 'dimensions)');
    console.log('--------------------------------------\n');

    const queryResponse = await container.items
        .query(querySpec)
        .fetchAll();

    const activityId = getQueryActivityId(queryResponse);
    if (activityId) {
        console.log('Query activity ID:', activityId);
    }

    const { resources, requestCharge } = queryResponse;
    printSearchResults(resources, requestCharge);
}

async function runMetricComparison(database: any, embedding: number[], safeEmbeddedField: string, data: any[]) {
    const rows: ComparisonRow[] = [];

    console.log('\nComparing distance metrics across DiskANN and QuantizedFlat');

    for (const algorithm of ALGORITHM_ORDER) {
        const algorithmConfig = algorithmConfigs[algorithm];
        const container = database.container(algorithmConfig.containerName);
        await container.read();
        await insertData(container, data);

        for (const metric of VALID_DISTANCE_FUNCTIONS) {
            const querySpec = buildVectorQuerySpec(embedding, safeEmbeddedField, metric, 2, 'SimilarityScore');
            const queryResponse = await container.items.query(querySpec).fetchAll();

            rows.push({
                algorithmName: algorithmConfig.algorithmName,
                metric,
                results: queryResponse.resources
            });
        }
    }

    printComparisonTable(rows);
}

async function main() {
    const { aiClient, dbClient } = getClientsPasswordless();

    try {
        // Validate algorithm selection
        if (!Object.keys(algorithmConfigs).includes(config.algorithm)) {
            throw new Error(`Invalid algorithm '${config.algorithm}'. Must be one of: ${Object.keys(algorithmConfigs).join(', ')}`);
        }

        // Validate distance function
        if (!VALID_DISTANCE_FUNCTIONS.includes(config.distanceFunction as (typeof VALID_DISTANCE_FUNCTIONS)[number])) {
            throw new Error(`Invalid distance function '${config.distanceFunction}'. Must be one of: ${VALID_DISTANCE_FUNCTIONS.join(', ')}`);
        }

        if (!aiClient) {
            throw new Error('Azure OpenAI client is not configured. Please check your environment variables.');
        }
        if (!dbClient) {
            throw new Error('Database client is not configured. Please check your environment variables.');
        }

        const algorithmConfig = algorithmConfigs[config.algorithm];
        const collectionName = algorithmConfig.containerName;

        try {
            const database = dbClient.database(config.dbName);
            console.log(`Connected to database: ${config.dbName}`);
            console.log(`\nVector Search Algorithm: ${algorithmConfig.algorithmName}`);
            console.log(`Distance Function: ${config.distanceFunction}`);
            if (config.compareMetrics) {
                console.log('Comparison Mode: metrics across DiskANN and QuantizedFlat');
            }

            const data = await readFileReturnJson(path.join(__dirname, "..", config.dataFile));

            const createEmbeddedForQueryResponse = await aiClient.embeddings.create({
                model: config.deployment,
                input: [config.query]
            });

            const safeEmbeddedField = validateFieldName(config.embeddedField);

            if (config.compareMetrics) {
                await runMetricComparison(database, createEmbeddedForQueryResponse.data[0].embedding, safeEmbeddedField, data);
            } else {
                const container = database.container(collectionName);
                console.log(`Connected to container: ${collectionName}`);
                await container.read();
                await insertData(container, data);
                await runSingleMetricQuery(container, createEmbeddedForQueryResponse.data[0].embedding, safeEmbeddedField, config.distanceFunction);
            }

            console.log('Vector search completed successfully!');
        } catch (error) {
            if ((error as any).code === 404) {
                throw new Error(`Container or database not found. Ensure database '${config.dbName}' and container '${collectionName}' exist before running this script.`);
            }
            throw error;
        }
    } catch (error) {
        console.error('App failed:', error);
        process.exitCode = 1;
    }
}

// Execute the main function
main().catch(error => {
    console.error('Unhandled error:', error);
    process.exitCode = 1;
});

File: nosql-vector-search-python/src/vector_search.py
"""Azure Cosmos DB NoSQL Vector Search — main entry point.

Loads hotel data, bulk-inserts into the selected container (DiskANN or
QuantizedFlat), generates a query embedding via Azure OpenAI, and
executes a VectorDistance() similarity search.
"""

import os
import sys
from pathlib import Path

from dotenv import load_dotenv

sys.path.insert(0, str(Path(__file__).parent))

from utils import (
    get_clients_passwordless,
    get_clients,
    insert_data,
    print_search_results,
    read_file_return_json,
    validate_field_name,
    get_query_activity_id,
)

# ---------------------------------------------------------------------------
# Load environment
# ---------------------------------------------------------------------------
load_dotenv()

ALGORITHM_CONFIGS: dict[str, dict[str, str]] = {
    "diskann": {
        "container_name": "hotels_diskann",
        "algorithm_name": "DiskANN",
    },
    "quantizedflat": {
        "container_name": "hotels_quantizedflat",
        "algorithm_name": "QuantizedFlat",
    },
}


VALID_DISTANCE_FUNCTIONS = ["cosine", "euclidean", "dotproduct"]


def _build_config() -> dict[str, str | int | bool]:
    """Build runtime configuration from environment variables."""
    distance_func = os.getenv("VECTOR_DISTANCE_FUNCTION", "cosine").strip().lower()
    if distance_func not in VALID_DISTANCE_FUNCTIONS:
        valid = ", ".join(VALID_DISTANCE_FUNCTIONS)
        raise ValueError(
            f"Invalid distance function '{distance_func}'. Must be one of: {valid}"
        )
    
    return {
        "query": "quintessential lodging near running trails, eateries, retail",
        "db_name": os.getenv("AZURE_COSMOSDB_DATABASENAME", "Hotels"),
        "algorithm": os.getenv("VECTOR_ALGORITHM", "diskann").strip().lower(),
        "data_file": os.getenv("DATA_FILE_WITH_VECTORS", "../data/HotelsData_toCosmosDB_Vector.json"),
        "embedded_field": os.getenv("EMBEDDED_FIELD", "DescriptionVector"),
        "embedding_dimensions": int(os.getenv("EMBEDDING_DIMENSIONS", "1536")),
        "deployment": os.getenv("AZURE_OPENAI_EMBEDDING_MODEL", "text-embedding-3-small"),
        "distance_function": distance_func,
        "compare_all_metrics": os.getenv("COMPARE_DISTANCE_METRICS", "false").lower() == "true",
    }


def main() -> None:
    """Run the vector search demonstration."""
    config = _build_config()

    # Try passwordless auth first, fall back to key-based
    clients = get_clients_passwordless()
    if not clients["ai_client"] or not clients["db_client"]:
        clients = get_clients()

    ai_client = clients["ai_client"]
    db_client = clients["db_client"]

    try:
        algorithm = config["algorithm"]
        if algorithm not in ALGORITHM_CONFIGS:
            valid = ", ".join(ALGORITHM_CONFIGS)
            raise ValueError(
                f"Invalid algorithm '{algorithm}'. Must be one of: {valid}"
            )

        if not ai_client:
            raise RuntimeError(
                "Azure OpenAI client is not configured. "
                "Please check your environment variables."
            )
        if not db_client:
            raise RuntimeError(
                "Cosmos DB client is not configured. "
                "Please check your environment variables."
            )

        algo_cfg = ALGORITHM_CONFIGS[algorithm]
        container_name = algo_cfg["container_name"]

        database = db_client.get_database_client(config["db_name"])
        print(f"Connected to database: {config['db_name']}")

        container = database.get_container_client(container_name)
        print(f"Connected to container: {container_name}")
        print(f"\n📊 Vector Search Algorithm: {algo_cfg['algorithm_name']}")

        # Verify the container exists
        try:
            container.read()
        except Exception as e:
            status_code = getattr(e, "status_code", None)
            if status_code == 404:
                raise RuntimeError(
                    f"Container or database not found. Ensure database "
                    f"'{config['db_name']}' and container '{container_name}' "
                    f"exist before running this script."
                ) from e
            raise

        data_path = Path(__file__).parent.parent / config["data_file"]
        data = read_file_return_json(str(data_path))
        insert_data(container, data)

        embedding_response = ai_client.embeddings.create(
            model=config["deployment"],
            input=[config["query"]],
        )
        query_embedding = embedding_response.data[0].embedding
        print(f"\n📊 Embedding generated: type={type(query_embedding)}, length={len(query_embedding)}, first_3_elements={query_embedding[:3]}")

        safe_field = validate_field_name(config["embedded_field"])
        
        # Run comparison if enabled, otherwise single metric
        if config["compare_all_metrics"]:
            _run_metric_comparison(
                container, safe_field, query_embedding, 
                config["query"], query_embedding
            )
        else:
            _run_single_metric_query(
                container, safe_field, query_embedding,
                config["distance_function"], config["query"]
            )

    except Exception as error:
        print(f"App failed: {error}", file=sys.stderr)
        sys.exit(1)


def _run_single_metric_query(
    container, safe_field: str, query_embedding: list, 
    distance_function: str, query_text: str
) -> None:
    """Execute vector search with a single distance metric."""
    print(f"📏 Distance Function: {distance_function}")
    
    query = (
        f'SELECT TOP 5 c.HotelName, c.Description, c.Rating, '
        f'VectorDistance(c.{safe_field}, @embedding, false, {{"distanceFunction": "{distance_function}"}}) AS SimilarityScore '
        f'FROM c '
        f'ORDER BY VectorDistance(c.{safe_field}, @embedding, false, {{"distanceFunction": "{distance_function}"}})'
    )

    print("\n--- Executing Vector Search Query ---")
    print(f"Query: {query}")
    print(
        f"Parameters: @embedding (vector with {len(query_embedding)} dimensions)"
    )
    print("--------------------------------------\n")

    results = list(
        container.query_items(
            query=query,
            parameters=[{"name": "@embedding", "value": query_embedding}],
            enable_cross_partition_query=True,
        )
    )

    # Extract diagnostics
    response_headers = container.client_connection.last_response_headers
    activity_id = get_query_activity_id(response_headers)
    if activity_id:
        print(f"Query activity ID: {activity_id}")

    request_charge_raw = response_headers.get("x-ms-request-charge", "0") if response_headers else "0"
    try:
        request_charge = float(request_charge_raw)
    except (ValueError, TypeError):
        request_charge = 0.0

    print_search_results(results, request_charge)


def _run_metric_comparison(
    container, safe_field: str, query_embedding: list,
    query_text: str, embedding: list
) -> None:
    """Execute vector search with all 3 distance metrics and display comparison."""
    print("📏 Comparing all distance functions: cosine, euclidean, dotproduct\n")
    
    metrics = ["cosine", "euclidean", "dotproduct"]
    metric_display_names = {"cosine": "Cosine", "euclidean": "Euclidean", "dotproduct": "DotProduct"}
    all_results = {}
    total_charge = 0.0
    
    for metric in metrics:
        display_name = metric_display_names[metric]
        print(f"\n--- {display_name} Distance Search ---")
        
        query = (
            f'SELECT TOP 5 c.HotelName, c.Description, c.Rating, '
            f'VectorDistance(c.{safe_field}, @embedding, false, {{"distanceFunction": "{metric}"}}) AS SimilarityScore '
            f'FROM c '
            f'ORDER BY VectorDistance(c.{safe_field}, @embedding, false, {{"distanceFunction": "{metric}"}})'
        )
        
        print(f"\n--- Query for {display_name} ---")
        print(f"Query: {query}")
        print(f"Embedding param type: {type(query_embedding)}, length: {len(query_embedding)}")
        
        results = list(
            container.query_items(
                query=query,
                parameters=[{"name": "@embedding", "value": query_embedding}],
                enable_cross_partition_query=True,
            )
        )
        
        # Extract diagnostics
        response_headers = container.client_connection.last_response_headers
        request_charge_raw = response_headers.get("x-ms-request-charge", "0") if response_headers else "0"
        try:
            request_charge = float(request_charge_raw)
        except (ValueError, TypeError):
            request_charge = 0.0
        
        total_charge += request_charge
        all_results[metric] = {
            "results": results,
            "charge": request_charge
        }
        
        print_search_results(results, request_charge)
    
    # Print comparison summary
    print("\n" + "="*80)
    print("📊 DISTANCE METRIC COMPARISON SUMMARY")
    print("="*80)
    
    for metric in metrics:
        display_name = metric_display_names[metric]
        charge = all_results[metric]["charge"]
        print(f"\n{display_name}:")
        print(f"  Request Charge: {charge:.2f} RUs")
        if all_results[metric]["results"]:
            top_result = all_results[metric]["results"][0]
            print(f"  Top Result: {top_result.get('HotelName', 'N/A')} (Score: {top_result.get('SimilarityScore', 'N/A'):.4f})")
    
    print(f"\nTotal Request Charge (all 3 metrics): {total_charge:.2f} RUs")
    print("="*80)


if __name__ == "__main__":
    main()
